### PR TITLE
Added storage for users parms as JSON

### PR DESCRIPTION
**Pull request by Criptos**
# Params as json store of `table_users`
On behalf of implementing this library as a micro service, we need to have  the minimal functional data set available. 
The user data is a variable, and the current solution of modifying the `table_user` to add more fields, makes the library hard to update, so, I added the parameter params and the table param_tables to solve this. 

Parameters can be added to `table_params` using the function addParam(), the arguments are:
*  $key the name of the key at the jsob object, cannot be empty and will be sanitized and lowerized
*  $name human name of the variable, suitable for UI, if empty, key will be used instead 
*  $required y/n, default y,  
*  $description description of the filed, human redeable, suitable for UI, can be empty
*  $filter, filter_var php to be applied, options: NONE,  FILTER_SANITIZE_EMAIL, FILTER_SANITIZE_ENCODED, FILTER_SANITIZE_NUMBER_FLOAT, FILTER_SANITIZE_NUMBER_INT, FILTER_SANITIZE_SPECIAL_CHARS, FILTER_SANITIZE_STRING, FILTER_SANITIZE_URL, default NONE 

There is also  the function removeParam to delete one param. 
 
Once  this fields are setup, if a param array is added to a parameter as required, and isn't set when the user is created or the params are modified, an error will be trow and user creation or data update, will not be done. 
 
__ To disable this functionality and keep the old functionality, just set config->table_params to '' (strlen == 0)  __

# Code cleanup
Also a little cleanup to the added code, instead of deleting the user if something failed a begin transaccion / rollback is used at the addUser funciton. 

# System errors added
(addParam) Key null or empty #17 
(addParam) filter not valir #18
(addParam) Key select query failed #19
(addParam) Duplicated key #20
(addParam) Insert failed to table_param #21

### CODE REVIEW

**Comment by KarelWintersky** [comment]
Interesting solution...

File: Auth.php [hunk 1]
```diff
@@ -678,34 +674,43 @@ protected function addUser($email, $password, $params = [], &$use_email_activati
         }
 
         $password = $this->getHash($password);
+        if(strlen($this->config->table_params)==0){
```
Comment (turbopixel) [inline Auth.php:677]: I prefer `empty(this->config->table_params)` (strlen is slow (performance))

**Comment by Criptos** [comment]
I'm at PHPAuth discord, if you have curiosity :)

File: Auth.php [hunk 2]
```diff
@@ -678,34 +674,43 @@ protected function addUser($email, $password, $params = [], &$use_email_activati
...
 
-            $query_prepared->execute([$uid]);
+        if (!$query_prepared->execute($bindParams)) {
+            $this->dbh->rollBack(); 
             $return['message'] = $this->__lang("system_error") . " #04";
-
             return $return;
         }
-
+        $this->dbh->commit();
```
Comment (turbopixel) [inline Auth.php:713]: How about checking the return value? In case of error the method returns false.

File: Auth.php [hunk 3]
```diff
@@ -1829,31 +1834,41 @@ public function updateUser($uid, $params)
 	
 	//unset uid which is set in getUser(). array generated in getUser() is now usable as parameter for updateUser()
 	unset($params['uid']);
+        if(strlen($this->config->tables_params)==0){
```
Comment (turbopixel) [inline Auth.php:1837]: I Prefer `empty()` (performance)

File: Auth.php [hunk 4]
```diff
@@ -1930,4 +1945,104 @@ public function cron()
...
+        
+        if(is_null($key)||empty($key)){
+            $return['message'] = $this->__lang("system_error" . "#17"); 
+            return $return ; 
+        }
+        if(!in_array($filter, $filterOptions)){
+            $return['message'] = $this->__lang("system_error" . "#18"); 
+            return $return ; 
+        }
+        if(is_null($name)||empty($name))
```
Comment (turbopixel) [inline Auth.php:1971]: We should use brackets :)

File: Auth.php [hunk 5]
```diff
@@ -1930,4 +1945,104 @@ public function cron()
...
+       * @returns array built from tables_params definition and sanitizing filter applied
+       */
+    private function appyFilterParams($params)
+    {
+        $jsonParams=$this->buildParams();
+        $finalParams=array();
+        foreach($jsonParams as $k=>$v){
+            if(strcmp($v['required'], 'y') == 0 && !isset($params[$k]))
+                throw new \Exception("${k} required but not set"); 
+            if(strcmp($v['filter'], 'NONE')==0)
```
Comment (turbopixel) [inline Auth.php:2040]: Please use brackets

**Review by turbopixel — COMMENTED** [review]
Some comments :)

**Comment by KarelWintersky** [comment]
> I'm at PHPAuth discord 

Where?

File: Auth.php (hunk 1 above)
Comment (Criptos) [reply Auth.php:677]: There was a reaaaaly strange situation when I used empty $this I kept getting affirmative assertion, so I ended with the sterlen is slow :(... I could change them to strcmp($this->config->table_params, '') maybe?

File: Auth.php (hunk 2 above)
Comment (Criptos) [reply Auth.php:713]: what method? that part of the function is only when the data is inserted... therefore the commit...

File: Auth.php (hunk 5 above)
Comment (Criptos) [reply Auth.php:2040]: you mean enclose { in throw }?

**Pull request closed without merging**